### PR TITLE
fix(compiler-cli): defer symbols only used in types

### LINKED ISSUE #55991: Angular @defer breaks when a deferred component is used as a GENERIC TYPE (not a value) of the viewChild
### Which @angular/* package(s) are the source of the bug?

core

### Is this a regression?

No

### Description

When using `viewChild` to query for a component wrapped with `@defer` and using the deferred component class as a generic type of the `viewChild`, eg `child = viewChild<ChildComponent>('someTplVariable')` the `@defer` behavior breaks and the deferred component bundle is not extracted as a lazy loaded bundle.

Steps to reproduce (also available inline in the Stackblitz)

  0. open the provided Stackblitz
  
  1. open demo in separate tab (so it's easy to see network tab traffic without Stackblitz specific things)
  2. after 3 seconds, no additional bundle will be loaded  after 3 seconds (network tab)

  3. remove the generic type from view child
  4. save and refresh the opened tab
  5. the child bundle will be loaded after 3 seconds (network tab)

Expected behavior

Using the deferred component class **only as a generic type** will work with `@defer`

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/view-child-generic-arg-breaks-defer

### Please provide the exception or error you saw

_No response_

### Please provide the environment you discovered this bug in (run `ng version`)

```
     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 17.3.7
Node: 18.18.0
Package Manager: npm 10.2.3
OS: linux x64

Angular: 17.3.9
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1703.7
@angular-devkit/build-angular   17.3.7
@angular-devkit/core            17.3.7
@angular-devkit/schematics      17.3.7
@angular/cli                    17.3.7
@schematics/angular             17.3.7
rxjs                            7.8.1
typescript                      5.3.3


```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by crisbeto**
Currently we don't defer any symbols that have references outside of the `import` statement and the `imports` array. This is a bit too aggressive, because it's possible that the symbol is only used for types (e.g. `viewChild<SomeCmp>('ref')`) which will be stripped when emitting to JS.

These changes expand the logic so that references inside type nodes aren't considered.

**Note:** one special case is when the symbol is used in constructor-based DI (e.g. `constructor(someCmp: SomeCmp)`, because these constructors will be compiled to `directiveInject` calls. We don't need to worry about them, because the compiler introduces an additional `import * as i1 from './some-cmp';` import that it uses to refer to the symbol.

Fixes #55991.

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/imports/src/deferred_symbol_tracker.ts [hunk 1]
```diff
@@ -171,7 +171,7 @@ export class DeferredSymbolTracker {
     }
 
     const symbolsMap = this.imports.get(importDecl)!;
-    for (const [symbol, refs] of symbolsMap) {
```
Comment (crisbeto) [inline packages/compiler-cli/src/ngtsc/imports/src/deferred_symbol_tracker.ts:174]: Unrelated change. I saw that we only used the type here.

File: packages/compiler-cli/test/ngtsc/defer_spec.ts [hunk 2]
```diff
@@ -79,12 +79,10 @@ runInEachFileSystem(() => {
       expect(jsContents).not.toContain('import { CmpA }');
     });
 
-    it(
```
Comment (crisbeto) [inline packages/compiler-cli/test/ngtsc/defer_spec.ts:82]: Unrelated change. I saw that the formatting was a bit weird here.

**Comment by tomastrajan** [comment]
Thanks for the great fix!

**Review by JoostK — APPROVED** [review]

**Comment by AndrewKushnir** [comment]
This PR was merged into the repository by commit 8d8c03abc40099da268d7301f029954f3e3f1c90.

The changes were merged into the following branches: main, 18.2.x
